### PR TITLE
Suggestions Api is created

### PR DESCRIPTION
**Pull request by sjha24**
1. Create Suggestion

### CODE REVIEW

File: src/main/java/com/neonlab/product/apis/DeleteSuggestionApi.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+
+
+@Service
+public class DeleteSuggestionApi {
+
+    @Autowired
+    SuggestionService suggestionService;
```
Comment (neon-lab-dev) [inline src/main/java/com/neonlab/product/apis/DeleteSuggestionApi.java]: Make it private

File: src/main/java/com/neonlab/product/service/SuggestionService.java [hunk 2]
```diff
@@ -29,6 +32,37 @@ public ApiOutput<SuggestionDto> createSuggestion(SuggestionDto suggestionDto) th
         return new ApiOutput<>(HttpStatus.OK.value(), "Thank You for Giving me Suggestion",suggestionDtos);
     }
 
+    public ApiOutput<List<SuggestionDto>> fetchSuggestion() throws InvalidInputException, ServerException {
```
Comment (neon-lab-dev) [inline src/main/java/com/neonlab/product/service/SuggestionService.java]: *Naming Convention*

- The method `fetchSuggention` is in *SuggetionService*.  
- Motive of giving the name is to communicate the operation it's doing. 
- Here, `fetchSuggestion` tells to get the list of suggestions. 
- The same thing can all be communicated by `suggestionService.fetch`

In an Office, a person's work is only to maintain the in and outs of employees in a record. 
If you want him to give the record, you will say `person.record` and not `person.inAndOutRecord`. But if you are asking him some other record that is out of his `concern` then you need to explicitly define which like `person.someOtherRecord`.

cc- @sjha24

Comment (neon-lab-dev) [reply src/main/java/com/neonlab/product/service/SuggestionService.java]: Service layer should not give ApiOut put . This is the concern of the Api layer and not service layer.

Give normal output. Attach that to the ApiOutput in the API layer.

File: src/main/java/com/neonlab/product/apis/SuggestionApi.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+
+
+
+@Service
+public class SuggestionApi {
+
+    @Autowired
+    SuggestionService suggestionService;
```
Comment (neon-lab-dev) [inline src/main/java/com/neonlab/product/apis/SuggestionApi.java]: Make these private

File: src/main/java/com/neonlab/product/controller/SuggestionController.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/v1/suggestion")
+public class SuggestionController {
+
+    private final SuggestionApi suggestionApi;
+
+    @PostMapping("/create")
+    public ApiOutput<SuggestionDto> createSuggestion(@RequestBody SuggestionDto suggestionDto){
+        return suggestionApi.createSuggestion(suggestionDto);
```
Comment (neon-lab-dev) [inline src/main/java/com/neonlab/product/controller/SuggestionController.java]: Refer to the `Naming Convention` comment.

File: src/main/java/com/neonlab/product/dtos/BoughtProductDetailsDto.java [hunk 5]
```diff
@@ -8,12 +10,17 @@
 
 
 @Data
```
Comment (neon-lab-dev) [inline src/main/java/com/neonlab/product/dtos/BoughtProductDetailsDto.java]: Add `@JsonIgnoreProperties(ignoreUnknown = true)`

File: src/main/java/com/neonlab/product/dtos/SuggestionDto.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package com.neonlab.product.dtos;
+import jakarta.validation.constraints.NotEmpty;
+import lombok.Data;
+
+
+
+@Data
+public class SuggestionDto {
+
+    String id;
```
Comment (neon-lab-dev) [inline src/main/java/com/neonlab/product/dtos/SuggestionDto.java]: add 
- createdAt
- createdBy
- ModifiedAt
- ModifiedBy

Comment (neon-lab-dev) [reply src/main/java/com/neonlab/product/dtos/SuggestionDto.java]: These should be the part of the response.

File: src/main/java/com/neonlab/product/entities/Suggestion.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+@EqualsAndHashCode(callSuper = true)
+@Entity
+@NoArgsConstructor
+@Data
+public class Suggestion extends Generic {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.UUID)
+    private String id;
+    private String comment;
```
Comment (neon-lab-dev) [inline src/main/java/com/neonlab/product/entities/Suggestion.java:24]: Use `@Column` make it as not nullable

File: src/main/java/com/neonlab/product/models/responses/PageableResponse.java [hunk 8]
```diff
@@ -11,7 +11,6 @@
 @NoArgsConstructor
 @JsonInclude(JsonInclude.Include.NON_NULL)
 public class PageableResponse<T> {
-
```
Comment (neon-lab-dev) [inline src/main/java/com/neonlab/product/models/responses/PageableResponse.java]: Why?

File: src/main/java/com/neonlab/product/service/OrderService.java [hunk 9]
```diff
@@ -53,69 +48,70 @@ public class OrderService {
     @Transactional
     public ApiOutput<?> createOrder(OrderDto orderDto) throws InvalidInputException, ServerException {
         Address shippingAddress = addressService.save(orderDto.getShippingInfo(), getUser());
-        List<BoughtProductDetailsDto> boughtProductDetailsDtoList = new ArrayList<>();
+        List<BoughtProductDetailsDto> boughtProductsDetailsList = new ArrayList<>();
```
Comment (neon-lab-dev) [inline src/main/java/com/neonlab/product/service/OrderService.java]: Learn using `var`

File: src/main/java/com/neonlab/product/service/SuggestionService.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+@Service
+public class SuggestionService {
+
+    @Autowired
+    SuggestionRepository suggestionRepository;
+    @Autowired
+    UserService userService;
+
+    public ApiOutput<SuggestionDto> createSuggestion(SuggestionDto suggestionDto) throws ServerException, InvalidInputException {
```
Comment (neon-lab-dev) [inline src/main/java/com/neonlab/product/service/SuggestionService.java]: ApiOut is a part of API layer and not service layer.

cc- @sjha24

**Review by neon-lab-dev — CHANGES_REQUESTED** [review]

File: src/main/java/com/neonlab/product/apis/SuggestionApi.java [hunk 11]
```diff
@@ -15,14 +15,14 @@
...
     @Autowired
-    ValidationUtils validationUtils;
+    private ValidationUtils validationUtils;
 
-    public ApiOutput<SuggestionDto> createSuggestion(SuggestionDto suggestionDto) {
+    public ApiOutput<SuggestionDto> create(SuggestionDto suggestionDto) {
         try {
             validationUtils.validation(suggestionDto,"suggestionDto");
-            return suggestionService.createSuggestion(suggestionDto);
+            return new ApiOutput<>(HttpStatus.OK.value(),"Thank Your for giving me suggestion" ,suggestionService.create(suggestionDto));
```
Comment (neon-lab-dev) [inline src/main/java/com/neonlab/product/apis/SuggestionApi.java]: Thank you for your suggestions

File: src/main/java/com/neonlab/product/apis/FetchSuggestionApi.java [hunk 12]
```diff
@@ -2,24 +2,21 @@
...
 
     @Autowired
-    SuggestionService suggestionService;
+    private SuggestionService suggestionService;
 
-    public ApiOutput<List<SuggestionDto>> fetchSuggestion() {
+    public ApiOutput<?> fetch() {
         try{
-            return suggestionService.fetchSuggestion();
+            return new ApiOutput<>(HttpStatus.OK.value(), "Your All Suggestion is below",suggestionService.fetch());
```
Comment (neon-lab-dev) [inline src/main/java/com/neonlab/product/apis/FetchSuggestionApi.java]: Suggestion List

File: src/main/java/com/neonlab/product/controller/SuggestionController.java [hunk 13]
```diff
@@ -20,17 +19,17 @@ public class SuggestionController {
     private final DeleteSuggestionApi deleteSuggestionApi;
 
     @PostMapping("/create")
```
Comment (neon-lab-dev) [inline src/main/java/com/neonlab/product/controller/SuggestionController.java:24]: Define roles permission

File: src/main/java/com/neonlab/product/controller/SuggestionController.java
Comment (neon-lab-dev) [inline src/main/java/com/neonlab/product/controller/SuggestionController.java:1]: Missing implementation for fetchall for admin only.

**Review by neon-lab-dev — CHANGES_REQUESTED** [review]

File: src/main/java/com/neonlab/product/apis/CancelOrderApi.java [hunk 14]
```diff
@@ -14,10 +14,10 @@ public class CancelOrderApi {
     @Autowired
     OrderService orderService;
 
-    public ApiOutput<?> cancelOrder(String orderId) {
+    public ApiOutput<?> cancelById(String orderId) {
```
Comment (ritik-agrawal) [inline src/main/java/com/neonlab/product/apis/CancelOrderApi.java:19]: Undo this previous was correct.

Comment (ritik-agrawal) [reply src/main/java/com/neonlab/product/apis/CancelOrderApi.java:19]: make it simply cancel

File: src/main/java/com/neonlab/product/dtos/SuggestionDto.java [hunk 15]
```diff
@@ -11,4 +13,8 @@ public class SuggestionDto {
 
     @NotEmpty(message = "Comment should not be empty")
     String comment;
+    Date createdAt;
```
Comment (ritik-agrawal) [inline src/main/java/com/neonlab/product/dtos/SuggestionDto.java]: private

Comment (ritik-agrawal) [reply src/main/java/com/neonlab/product/dtos/SuggestionDto.java]: comment v private hoga

**Review by ritik-agrawal — DISMISSED** [review]

**Review by neon-lab-dev — APPROVED** [review]

**Pull request merged**